<a href="https://colab.research.google.com/github/xxrdfixx/AI_Applications/blob/main/Week_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import tensorflow as tf
import matplotlib.pyplot as plt

In [3]:
#Load Data
mnist = tf.keras.datasets.mnist

#Split data into train and test
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

#Normalize images
train_images=train_images/255.0
test_images=test_images/255.0

11490434/11490434 [==============================] - 1s 0us/step


In [4]:
#2.1 Advanced Activity: Implement a CNN with Dropout and Batch Normalization
def evaluate_cnn():
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation='relu'),
        tf.keras.layers.BatchNormalization(),  # Batch Normalization layer
        tf.keras.layers.Dropout(0.5),  # Dropout layer with a dropout rate of 0.5
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with CNN, Dropout, and Batch Normalization: {test_acc}")

    return model

cnn_model = evaluate_cnn()


# Dropout is a regularization technique used during training to prevent overfitting.
# It works by randomly setting a fraction of the input units to a layer to zero during each forward and backward pass.
# During training, at each iteration, dropout randomly "drops out" (sets to zero) a fraction of the neurons in the network.

# Batch Normalization is used to improve the stability and convergence of neural networks during training by normalizing the inputs to each layer.
# Batch Normalization computes the mean and standard deviation of the input data within a mini-batch during training.
# It then scales and shifts the inputs to ensure that they have a specified mean and standard deviation (typically close to 0 and 1).
# This normalization is applied before the activation function of each layer.

Epoch 1/10
1875/1875 [==============================] - 42s 21ms/step - loss: 0.1731 - accuracy: 0.9475
Epoch 2/10
1875/1875 [==============================] - 40s 22ms/step - loss: 0.0836 - accuracy: 0.9741
Epoch 3/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0669 - accuracy: 0.9793
Epoch 4/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0528 - accuracy: 0.9833
Epoch 5/10
1875/1875 [==============================] - 42s 23ms/step - loss: 0.0440 - accuracy: 0.9859
Epoch 6/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0412 - accuracy: 0.9873
Epoch 7/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0334 - accuracy: 0.9892
Epoch 8/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0333 - accuracy: 0.9891
Epoch 9/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0275 - accuracy: 0.9907
Epoch 10/10
313/313 [==============================] - 2s 6ms/st

In [5]:
# 2.2 Activity 2: Experiment with Different Activation Functions
def evaluate_cnn_with_activation(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {activation_function} activation: {test_acc}")

    return model

for activation_function in ['relu', 'sigmoid', 'tanh', 'elu']:
    print(f"Experimenting with {activation_function} activation:")
    evaluate_cnn_with_activation(activation_function)

# Relu had the lowest accuracy, followed by sigmoid while elu had the highest followed by tahn.

Experimenting with relu activation:
Epoch 1/10
1875/1875 [==============================] - 47s 25ms/step - loss: 0.1534 - accuracy: 0.9545
Epoch 2/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0521 - accuracy: 0.9837
Epoch 3/10
1875/1875 [==============================] - 43s 23ms/step - loss: 0.0341 - accuracy: 0.9891
Epoch 4/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0224 - accuracy: 0.9931
Epoch 5/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0153 - accuracy: 0.9949
Epoch 6/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0107 - accuracy: 0.9967
Epoch 7/10
1875/1875 [==============================] - 41s 22ms/step - loss: 0.0086 - accuracy: 0.9969
Epoch 8/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0063 - accuracy: 0.9980
Epoch 9/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0048 - accuracy: 0.9985
Epoch 10/10
313/313 [=======

In [6]:
# 2.3 Activity 3: Add More Convolutional Layers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),  # Additional Conv2D layer
        tf.keras.layers.MaxPooling2D((2, 2)),

        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 56s 29ms/step - loss: 0.1295 - accuracy: 0.9604
Epoch 2/10
1875/1875 [==============================] - 57s 30ms/step - loss: 0.0440 - accuracy: 0.9868
Epoch 3/10
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0297 - accuracy: 0.9909
Epoch 4/10
1875/1875 [==============================] - 55s 30ms/step - loss: 0.0219 - accuracy: 0.9928
Epoch 5/10
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0156 - accuracy: 0.9949
Epoch 6/10
1875/1875 [==============================] - 56s 30ms/step - loss: 0.0128 - accuracy: 0.9958
Epoch 7/10
1875/1875 [==============================] - 55s 30ms/step - loss: 0.0105 - accuracy: 0.9965
Epoch 8/10
1875/1875 [==============================] - 55s 29ms/step - loss: 0.0084 - accuracy: 0.9972
Epoch 9/10
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0068 - accuracy: 0.9978
Epoch 10/10
313/313 [==============================] - 3s 9ms/st

In [8]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='SGD', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 53s 28ms/step - loss: 0.4696 - accuracy: 0.8726
Epoch 2/10
1875/1875 [==============================] - 33s 18ms/step - loss: 0.2169 - accuracy: 0.9360
Epoch 3/10
1875/1875 [==============================] - 34s 18ms/step - loss: 0.1655 - accuracy: 0.9512
Epoch 4/10
1875/1875 [==============================] - 35s 19ms/step - loss: 0.1347 - accuracy: 0.9605
Epoch 5/10
1875/1875 [==============================] - 36s 19ms/step - loss: 0.1129 - accuracy: 0.9667
Epoch 6/10
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0981 - accuracy: 0.9711
Epoch 7/10
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0868 - accuracy: 0.9745
Epoch 8/10
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0780 - accuracy: 0.9768
Epoch 9/10
1875/1875 [==============================] - 35s 19ms/step - loss: 0.0703 - accuracy: 0.9790
Epoch 10/10
313/313 [==============================] - 2s 7ms/st

In [9]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='RMSprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.1514 - accuracy: 0.9546
Epoch 2/10
1875/1875 [==============================] - 37s 20ms/step - loss: 0.0522 - accuracy: 0.9845
Epoch 3/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0347 - accuracy: 0.9897
Epoch 4/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.0245 - accuracy: 0.9927
Epoch 5/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.0189 - accuracy: 0.9944
Epoch 6/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.0127 - accuracy: 0.9962
Epoch 7/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0091 - accuracy: 0.9974
Epoch 8/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.0065 - accuracy: 0.9981
Epoch 9/10
1875/1875 [==============================] - 38s 20ms/step - loss: 0.0049 - accuracy: 0.9987
Epoch 10/10
313/313 [==============================] - 2s 6ms/st

In [10]:
# 2.4 Activity 4: Experiment with Different Optimizers
def evaluate_cnn_with_layer(activation_function):
    model = tf.keras.Sequential([
        tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        tf.keras.layers.MaxPooling2D((2, 2)),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(128, activation=activation_function),  # Experiment with different activations here
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=10)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy: {test_acc}")

    return model

cnn_model = evaluate_cnn_with_layer('relu')

Epoch 1/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.1504 - accuracy: 0.9565
Epoch 2/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0525 - accuracy: 0.9835
Epoch 3/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0333 - accuracy: 0.9895
Epoch 4/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0216 - accuracy: 0.9929
Epoch 5/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0144 - accuracy: 0.9953
Epoch 6/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0105 - accuracy: 0.9966
Epoch 7/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0076 - accuracy: 0.9975
Epoch 8/10
1875/1875 [==============================] - 39s 21ms/step - loss: 0.0066 - accuracy: 0.9979
Epoch 9/10
1875/1875 [==============================] - 40s 21ms/step - loss: 0.0050 - accuracy: 0.9985
Epoch 10/10
313/313 [==============================] - 2s 7ms/st

### **RMSprop had the best performance at 0.985 accuracy, followed by adam at 0.9825 and lastly, relu at 0.976.**